File: Pinball App 2/Pinball App 2/settings/SettingsScreen.swift
import SwiftUI
import Combine

enum SettingsRoute: Hashable {
    case addManufacturer
    case addVenue
    case addTournament
}

@MainActor
final class SettingsViewModel: ObservableObject {
    @Published private(set) var manufacturers: [PinballCatalogManufacturerOption] = []
    @Published private(set) var importedSources: [PinballImportedSourceRecord] = []
    @Published private(set) var sourceState: PinballLibrarySourceState = .empty
    @Published private(set) var isLoading = false
    @Published private(set) var isRefreshingHostedData = false
    @Published private(set) var hostedDataStatusMessage: String?
    @Published private(set) var hostedDataStatusIsError = false
    @Published private(set) var isClearingCache = false
    @Published private(set) var cacheStatusMessage: String?
    @Published private(set) var cacheStatusIsError = false
    @Published var errorMessage: String?

    private var didLoad = false
    let builtinSources: [PinballLibrarySource] = builtinVenueSources()

    private func applySnapshot(_ snapshot: SettingsDataSnapshot) {
        manufacturers = snapshot.manufacturers
        importedSources = snapshot.importedSources
        sourceState = snapshot.sourceState
    }

    private func applySourceSnapshot(_ snapshot: SettingsSourceSnapshot) {
        importedSources = snapshot.importedSources
        sourceState = snapshot.sourceState
    }

    func loadIfNeeded() async {
        guard !didLoad else { return }
        didLoad = true
        await refresh()
    }

    func refresh() async {
        isLoading = true
        defer { isLoading = false }

        do {
            applySnapshot(try await loadSettingsDataSnapshot())
            errorMessage = nil
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func forceRefreshHostedLibraryData() async {
        guard !isRefreshingHostedData else { return }
        isRefreshingHostedData = true
        hostedDataStatusMessage = nil
        hostedDataStatusIsError = false
        defer { isRefreshingHostedData = false }

        do {
            applySnapshot(try await forceRefreshHostedSettingsData())
            hostedDataStatusMessage = "Pinball data refreshed from pillyliu.com."
            hostedDataStatusIsError = false
            postPinballLibrarySourcesDidChange()
        } catch {
            hostedDataStatusMessage = "Hosted data refresh failed: \(error.localizedDescription)"
            hostedDataStatusIsError = true
        }
    }

    func clearCachedData() async {
        guard !isClearingCache else { return }
        isClearingCache = true
        cacheStatusMessage = nil
        cacheStatusIsError = false
        defer { isClearingCache = false }

        do {
            try await clearAppRuntimeCaches()
            cacheStatusMessage = "Cached data cleared. Hosted data will refetch as screens reload."
            cacheStatusIsError = false
        } catch {
            cacheStatusMessage = "Cache clear failed: \(error.localizedDescription)"
            cacheStatusIsError = true
        }
    }

    func isEnabled(_ sourceID: String, builtin: Bool) -> Bool {
        builtin || sourceState.enabledSourceIDs.contains(sourceID)
    }

    func isPinned(_ sourceID: String) -> Bool {
        sourceState.pinnedSourceIDs.contains(sourceID)
    }

    func toggleEnabled(_ sourceID: String, builtin: Bool, isOn: Bool) {
        guard !builtin else { return }
        PinballLibrarySourceStateStore.setEnabled(sourceID: sourceID, isEnabled: isOn)
        sourceState = PinballLibrarySourceStateStore.load()
        postPinballLibrarySourcesDidChange()
    }

    func togglePinned(_ sourceID: String, isOn: Bool) {
        let success = PinballLibrarySourceStateStore.setPinned(sourceID: sourceID, isPinned: isOn)
        if !success {
            errorMessage = "Pinned sources are limited to \(PinballLibrarySourceStateStore.maxPinnedSources)."
        }
        sourceState = PinballLibrarySourceStateStore.load()
        postPinballLibrarySourcesDidChange()
    }

    func addManufacturer(_ manufacturer: PinballCatalogManufacturerOption) {
        applySourceSnapshot(addManufacturerSource(manufacturer))
        postPinballLibrarySourcesDidChange()
    }

    func importVenue(result: PinballLibraryVenueSearchResult, machineIDs: [String], searchQuery: String, radiusMiles: Int) {
        applySourceSnapshot(
            addVenueSource(
                result: result,
                machineIDs: machineIDs,
                searchQuery: searchQuery,
                radiusMiles: radiusMiles
            )
        )
        postPinballLibrarySourcesDidChange()
    }

    func importTournament(result: MatchPlayTournamentImportResult) {
        applySourceSnapshot(addTournamentSource(result))
        postPinballLibrarySourcesDidChange()
    }

    func removeImportedSource(_ sourceID: String) {
        applySourceSnapshot(removeSettingsSource(sourceID))
        postPinballLibrarySourcesDidChange()
    }

    func refreshVenue(_ source: PinballImportedSourceRecord) async {
        guard source.type == .venue else { return }
        do {
            applySourceSnapshot(try await refreshVenueSource(source))
            postPinballLibrarySourcesDidChange()
        } catch {
            errorMessage = "Venue refresh failed: \(error.localizedDescription)"
        }
    }

    func refreshTournament(_ source: PinballImportedSourceRecord) async {
        guard source.type == .tournament else { return }
        do {
            applySourceSnapshot(try await refreshTournamentSource(source))
            postPinballLibrarySourcesDidChange()
        } catch {
            errorMessage = "Tournament refresh failed: \(error.localizedDescription)"
        }
    }
}

struct SettingsScreen: View {
    @StateObject private var viewModel = SettingsViewModel()
    @State private var navigationPath: [SettingsRoute] = []
    @AppStorage(LPLNamePrivacySettings.fullNameAccessUnlockedDefaultsKey) private var lplFullNameAccessUnlocked = false
    @AppStorage(LPLNamePrivacySettings.showFullLastNameDefaultsKey) private var showFullLPLLastNames = false
    @AppStorage("app-intro-show-on-next-launch") private var appIntroShowOnNextLaunch = false
    @State private var lplNamePassword: String = ""
    @State private var lplNamePrivacyError: String?
    @State private var introOverlayToggleMessage: String?
    @State private var introOverlayToggleMessageTask: Task<Void, Never>?

    var body: some View {
        NavigationStack(path: $navigationPath) {
            AppScreen {
                ScrollView {
                    SettingsHomeContent(
                        viewModel: viewModel,
                        navigationPath: $navigationPath,
                        lplFullNameAccessUnlocked: $lplFullNameAccessUnlocked,
                        showFullLPLLastNames: $showFullLPLLastNames,
                        lplNamePassword: $lplNamePassword,
                        lplNamePrivacyError: $lplNamePrivacyError,
                        onToggleIntroOverlayForNextLaunch: toggleIntroOverlayForNextLaunch
                    )
                    .padding(.horizontal, 14)
                    .padding(.vertical, 10)
                }
            }
            .overlay(alignment: .top) {
                if let introOverlayToggleMessage {
                    AppSuccessBanner(text: introOverlayToggleMessage)
                        .padding(.top, 4)
                        .transition(.move(edge: .top).combined(with: .opacity))
                }
            }
            .animation(.easeInOut(duration: 0.25), value: introOverlayToggleMessage)
            .navigationTitle("Settings")
            .navigationBarTitleDisplayMode(.inline)
            .task {
                await viewModel.loadIfNeeded()
            }
            .onReceive(NotificationCenter.default.publisher(for: .pinballLibrarySourcesDidChange)) { _ in
                Task { await viewModel.refresh() }
            }
            .onDisappear {
                introOverlayToggleMessageTask?.cancel()
                introOverlayToggleMessageTask = nil
            }
            .navigationDestination(for: SettingsRoute.self) { route in
                settingsRouteDestination(route: route, viewModel: viewModel)
            }
        }
    }

    private func toggleIntroOverlayForNextLaunch() {
        appIntroShowOnNextLaunch.toggle()
        introOverlayToggleMessage = appIntroShowOnNextLaunch
            ? "Intro enabled for next launch"
            : "Intro disabled for next launch"

        introOverlayToggleMessageTask?.cancel()
        let message = introOverlayToggleMessage
        introOverlayToggleMessageTask = Task { @MainActor in
            try? await Task.sleep(for: .seconds(1.2))
            guard !Task.isCancelled, introOverlayToggleMessage == message else { return }
            introOverlayToggleMessage = nil
            introOverlayToggleMessageTask = nil
        }
    }
}

File: Pinball App Android/app/src/main/java/com/pillyliu/pinprofandroid/settings/SettingsScreen.kt
package com.pillyliu.pinprofandroid.settings

import androidx.compose.animation.AnimatedVisibility
import androidx.compose.animation.core.tween
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.padding
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.Alignment
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.unit.dp
import com.pillyliu.pinprofandroid.data.toggleShowAppIntroOverlayOnNextLaunch
import com.pillyliu.pinprofandroid.library.CatalogManufacturerOption
import com.pillyliu.pinprofandroid.library.LibrarySource
import com.pillyliu.pinprofandroid.library.LibrarySourceEvents
import com.pillyliu.pinprofandroid.library.LibrarySourceStateStore
import com.pillyliu.pinprofandroid.library.LibrarySourceType
import com.pillyliu.pinprofandroid.library.builtinVenueSources
import com.pillyliu.pinprofandroid.ui.AppSuccessBanner
import com.pillyliu.pinprofandroid.ui.AppScreen
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

sealed interface SettingsRoute {
    data object Home : SettingsRoute
    data object AddManufacturer : SettingsRoute
    data object AddVenue : SettingsRoute
    data object AddTournament : SettingsRoute
}

@Composable
internal fun SettingsScreen(contentPadding: PaddingValues) {
    val context = LocalContext.current
    val scope = rememberCoroutineScope()
    val state = rememberSettingsScreenState(context)
    var introOverlayToggleMessage by remember { mutableStateOf<String?>(null) }
    val builtinSources = remember {
        builtinVenueSources()
    }
    val sourceVersion by LibrarySourceEvents.version.collectAsState()

    LaunchedEffect(Unit) {
        state.reload()
    }
    LaunchedEffect(sourceVersion) {
        if (sourceVersion != 0L) {
            state.reload()
        }
    }
    LaunchedEffect(introOverlayToggleMessage) {
        if (introOverlayToggleMessage != null) {
            delay(1_200L)
            introOverlayToggleMessage = null
        }
    }

    when (state.route) {
        SettingsRoute.AddManufacturer -> {
            AddManufacturerScreen(
                contentPadding = contentPadding,
                manufacturers = state.manufacturers,
                onBack = { state.route = SettingsRoute.Home },
                onAdd = { manufacturer ->
                    state.applySourceSnapshot(addManufacturerSource(context, manufacturer))
                    state.route = SettingsRoute.Home
                },
            )
            return
        }

        SettingsRoute.AddVenue -> {
            AddVenueScreen(
                contentPadding = contentPadding,
                onBack = { state.route = SettingsRoute.Home },
                onImport = { result, machineIds, query, radiusMiles ->
                    state.applySourceSnapshot(addVenueSource(context, result, machineIds, query, radiusMiles))
                    state.route = SettingsRoute.Home
                },
            )
            return
        }

        SettingsRoute.AddTournament -> {
            AddTournamentScreen(
                contentPadding = contentPadding,
                onBack = { state.route = SettingsRoute.Home },
                onImport = { result ->
                    state.applySourceSnapshot(addTournamentSource(context, result))
                    state.route = SettingsRoute.Home
                },
            )
            return
        }

        SettingsRoute.Home -> Unit
    }

    AppScreen(contentPadding) {
        Box(modifier = Modifier.fillMaxSize()) {
            SettingsHomeContent(
                builtinSources = builtinSources,
                manufacturers = state.manufacturers,
                importedSources = state.importedSources,
                sourceState = state.sourceState,
                loading = state.loading,
                error = state.error,
                refreshingHostedData = state.refreshingHostedData,
                hostedDataStatusMessage = state.hostedDataStatusMessage,
                hostedDataStatusIsError = state.hostedDataStatusIsError,
                clearingCache = state.clearingCache,
                cacheStatusMessage = state.cacheStatusMessage,
                cacheStatusIsError = state.cacheStatusIsError,
                onOpenAddManufacturer = { state.route = SettingsRoute.AddManufacturer },
                onOpenAddVenue = { state.route = SettingsRoute.AddVenue },
                onOpenAddTournament = { state.route = SettingsRoute.AddTournament },
                onToggleEnabled = { sourceId, isEnabled ->
                    LibrarySourceStateStore.setEnabled(context, sourceId, isEnabled)
                    state.afterSourceMutation()
                },
                onTogglePinned = { sourceId, isPinned ->
                    if (LibrarySourceStateStore.setPinned(context, sourceId, isPinned)) {
                        state.afterSourceMutation()
                    } else {
                        state.error = "Pinned sources are limited to ${LibrarySourceStateStore.MAX_PINNED_SOURCES}."
                    }
                },
                onRefreshSource = { source ->
                    scope.launch {
                        when (source.type) {
                            LibrarySourceType.VENUE -> {
                                runCatching { refreshVenueSource(context, source) }
                                    .onSuccess(state::applySourceSnapshot)
                                    .onFailure {
                                        state.error = "Venue refresh failed: ${it.message ?: "Unknown error"}"
                                    }
                            }

                            LibrarySourceType.TOURNAMENT -> {
                                runCatching { refreshTournamentSource(context, source) }
                                    .onSuccess(state::applySourceSnapshot)
                                    .onFailure {
                                        state.error = "Tournament refresh failed: ${it.message ?: "Unknown error"}"
                                    }
                            }

                            else -> Unit
                        }
                    }
                },
                onDeleteSource = { sourceId ->
                    state.applySourceSnapshot(removeSettingsSource(context, sourceId))
                },
                onRefreshHostedData = {
                    scope.launch { state.refreshHostedLibraryData() }
                },
                onClearCache = {
                    scope.launch { state.clearCachedData() }
                },
                onToggleIntroOverlayForNextLaunch = {
                    val enabled = toggleShowAppIntroOverlayOnNextLaunch(context)
                    introOverlayToggleMessage = if (enabled) {
                        "Intro enabled for next launch"
                    } else {
                        "Intro disabled for next launch"
                    }
                },
            )

            AnimatedVisibility(
                visible = introOverlayToggleMessage != null,
                modifier = Modifier
                    .align(Alignment.TopCenter)
                    .padding(top = 4.dp),
                enter = fadeIn(animationSpec = tween(durationMillis = 250)),
                exit = fadeOut(animationSpec = tween(durationMillis = 250)),
            ) {
                introOverlayToggleMessage?.let { message ->
                    AppSuccessBanner(text = message)
                }
            }
        }
    }
}
